### ISSUE TITLE
DNS resolution is being cache in services

### ISSUE THREAD

**Issue by jalberto** [kontena/kontena#1852]
stackA expose a service to be used by stackB, stackB try to connect to stackA.grid.kontena.local, but if instances in stackA change IPs for any reason StackB continue resolving to old IP.

Restarting docker server in stackA nodes solves the problem

**Comment by SpComb** [comment]
The service container addresses for `stackA.grid.kontena.local` should be updated by the Kontena agent and Weave as any stackA services are re-deployed.

What do you mean by restarting the Docker server in stackA nodes? Do you mean `systemctl restart docker`, or `docker restart stackB.service-X`?

If you're restarting the connecting stackB service container, then it sounds like the application running there is caching the DNS lookup results at startup, and re-using the stale IP address instead of re-resolving it each time it tries to connect, which is unfortunately a somewhat common issue with some applications.

**Comment by jalberto** [comment]
I mean Docker itself.

**Comment by jnummelin** [comment]
But restarting docker itself means also restarting the application. Unless you've configured [live-restore](https://docs.docker.com/engine/admin/live-restore/) functionality on docker engine level.

**Comment by jalberto** [comment]
@jnummelin yes, it restarts everything. What about configuring the docker daemon to have a TTL=0 or similar?

**Comment by jnummelin** [comment]
As @SpComb mentioned, the most probable cacher in this case is the application itself so setting any TTL would no have effect.

What is the client side app where you see this issue?

**Comment by jalberto** [comment]
Elasticsearch (with LS_JAVA_OPTS: "-Dnetworkaddress.cache.ttl=1") & rails

**Comment by SpComb** [comment]
The DNS TTL shouldn't be relevant, because the `/etc/resolv.conf` in the Docker containers will have every application host lookup go directly to the local weave-dns server.
